spec/lib/stellar/sep10_spec.rb

```diff
@@ -0,0 +1,604 @@
+        {"key" => client_kp_c.address, "weight" => 4},
+      ]
+
+      signers_found = sep10.verify_challenge_tx_threshold(
+        challenge_xdr: challenge_tx,
+        server: server_kp,
+        threshold: 7,
+        signers: signers,
+      )
+      expect(signers_found).to eql(signers)
+    end
+
+    it "raises error when signers don't meet threshold" do
+      server_kp = Stellar::KeyPair.random
+      client_kp_a = Stellar::KeyPair.random
+      client_kp_b = Stellar::KeyPair.random
+      client_kp_c = Stellar::KeyPair.random
+      timeout = 600
+      anchor_name = "SDF"
+
+      challenge = sep10.build_challenge_tx(
+        server: server_kp,
+        client: client_kp_a,
+        anchor_name: anchor_name,
+        timeout: timeout,
+      )
+
+      transaction = Stellar::TransactionEnvelope.from_xdr(challenge, "base64")
+      transaction.signatures.push(transaction.tx.sign_decorated(client_kp_a))
+      challenge_tx = transaction.to_xdr(:base64)
+
+      signers = Set[
+        {"key" => client_kp_a.address, "weight" => 1},
+        {"key" => client_kp_b.address, "weight" => 2},
+        {"key" => client_kp_c.address, "weight" => 4},
+      ]
+
+      expect {
+        sep10.verify_challenge_tx_threshold(
+          challenge_xdr: challenge_tx,
+          server: server_kp,
+          threshold: 7,
+          signers: signers,
+        )
+      }.to raise_error(
+        Stellar::InvalidSep10ChallengeError,
+        "signers with weight %d do not meet threshold %d." % [1, 7]
+      )
+    end
+  end
+
+  describe "#verify_challenge_tx" do
+    it "verifies proper challenge transaction" do
+      server_kp = Stellar::KeyPair.random
+      client_kp = Stellar::KeyPair.random
+      timeout = 600
+      anchor_name = "SDF"
+
+      challenge = sep10.build_challenge_tx(
+        server: server_kp,
+        client: client_kp,
+        anchor_name: anchor_name,
+        timeout: timeout,
+      )
+
+      envelope = Stellar::TransactionEnvelope.from_xdr(challenge, "base64")
+      envelope.signatures.push(envelope.tx.sign_decorated(client_kp))
+      challenge_tx = envelope.to_xdr(:base64)
+
+      sep10.verify_challenge_tx(
+        challenge_xdr: challenge_tx, 
+        server: server_kp
+      )
+    end
+
+    it "raises not signed by client" do
+      server_kp = Stellar::KeyPair.random
+      client_kp = Stellar::KeyPair.random
+      timeout = 600
+      anchor_name = "SDF"
+
+      challenge = sep10.build_challenge_tx(
+        server: server_kp,
+        client: client_kp,
+        anchor_name: anchor_name,
+        timeout: timeout,
+      )
+
+      expect {
+        sep10.verify_challenge_tx(
+          challenge_xdr: challenge, 
+          server: server_kp
+        )
+      }.to raise_error(
+        Stellar::InvalidSep10ChallengeError,
+        "Transaction not signed by client: %s" % [client_kp.address]
+      )
+    end
+  end
+
+  describe "#verify_challenge_tx_signers" do
```

Comment:
Could we add tests for these use cases?

- [ ] include only the server in the list of signers to verify (error case)
- [ ] include the server in the list of signers along with client signers to verify (valid case)
- [ ] signed by server and clients all signatures used but some signers aren't used (valid case) ([Go SDK ref](https://github.com/stellar/go/blob/edbd8a0ac1d16206ce22f3ebdf00706b26657b46/txnbuild/transaction_test.go#L1675-L1713))
- [ ] duplicate signers
- [ ] duplicate signatures
- [ ] signed by server and clients all signatures used but some PreauthTxHash (T signer) and XHash (X signer) signers ignored (valid case) ([Go SDK ref](https://github.com/stellar/go/blob/edbd8a0ac1d16206ce22f3ebdf00706b26657b46/txnbuild/transaction_test.go#L1715-L1759))
- [ ] signatures unrecognized and can't be mapped to a signer (error case) ([Go SDK ref](https://github.com/stellar/go/blob/edbd8a0ac1d16206ce22f3ebdf00706b26657b46/txnbuild/transaction_test.go#L1796-L1828))
- [ ] no signatures (error case)